api/lib/domain/services/schooling-registrations-xml-service.js

```diff
@@ -1,91 +1,147 @@
+
+    const streamerToParseOrganizationUAI = new saxPath.SaXPath(saxParser, NODE_ORGANIZATION_UAI);
+    const streamerToParseSchoolingRegistrations = new saxPath.SaXPath(saxParser, NODES_SCHOOLING_REGISTRATIONS);
+
+    const mapSchoolingRegistrationsByStudentId = new Map();
+    const nationalStudentIds = [];
+
+    streamerToParseOrganizationUAI.on('match', (xmlFormat) => {
+      parseString(xmlFormat, function(err, jsFormat) {
+        if (jsFormat.PARAMETRES) {
+          const UAIFromSIECLE = _getValueFromParsedElement(jsFormat.PARAMETRES.UAJ);
+          if (UAIFromSIECLE !== organization.externalId) {
+            reject(new FileValidationError(UAI_SIECLE_FILE_NOT_MATCH_ORGANIZATION_UAI));
+          }
+        }
+      });
     });
 
-  return { UAIFromSIECLE, resultFromExtraction };
-}
+    streamerToParseSchoolingRegistrations.on('match', (xmlFormat) => {
+      if (xmlFormat.startsWith(ELEVE_ELEMENT) || xmlFormat.startsWith(STRUCTURE_ELEVE_ELEMENT)) {
+        parseString(xmlFormat, function(err, jsFormat) {
+
+          if (jsFormat.ELEVE) {
+            const isStudentNotLeftSchoolingRegistration = isEmpty(jsFormat.ELEVE.DATE_SORTIE);
+            const isStudentNotYetArrivedSchoolingRegistration = !isEmpty(jsFormat.ELEVE.ID_NATIONAL);
+            const isStudentNotDuplicatedInTheSIECLEFile = !mapSchoolingRegistrationsByStudentId.has(jsFormat.ELEVE.$.ELEVE_ID);
+
+            if (isStudentNotLeftSchoolingRegistration && isStudentNotYetArrivedSchoolingRegistration && isStudentNotDuplicatedInTheSIECLEFile) {
+              const nationalStudentId = _getValueFromParsedElement(jsFormat.ELEVE.ID_NATIONAL);
+              if (nationalStudentId && nationalStudentIds.indexOf(nationalStudentId) !== -1) {
+                reject(new SameNationalStudentIdInFileError(nationalStudentId));
+              }
+              nationalStudentIds.push(nationalStudentId);
+              mapSchoolingRegistrationsByStudentId.set(jsFormat.ELEVE.$.ELEVE_ID, _mapStudentInformationToSchoolingRegistration(jsFormat));
+            }
+          }
+
+          if (jsFormat.STRUCTURES_ELEVE && mapSchoolingRegistrationsByStudentId.has(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID)) {
+            const currentStudent = mapSchoolingRegistrationsByStudentId.get(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID);
+            const structureElement = jsFormat.STRUCTURES_ELEVE.STRUCTURE;
+            if (!isEmpty(structureElement)) {
+              structureElement.forEach((structure) => {
+                if (structure.TYPE_STRUCTURE[0] === DIVISION && structure.CODE_STRUCTURE[0] !== 'Inactifs') {
+                  currentStudent.division = structure.CODE_STRUCTURE[0];
+                }
+              });
+
+              if (currentStudent.division === undefined) {
+                mapSchoolingRegistrationsByStudentId.delete(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID);
+              }
+            }
```

Comment:
Lodash pourrait t'éviter de vérifier que les collections ne sont pas vides, et donc réduire le nesting d'un cran
```suggestion
         const structureElement = jsFormat.STRUCTURES_ELEVE.STRUCTURE;
         _.each(structureElement, (structure) => {
                if (structure.TYPE_STRUCTURE[0] === DIVISION && structure.CODE_STRUCTURE[0] !== 'Inactifs') {
                  currentStudent.division = structure.CODE_STRUCTURE[0];
                }
         });
        if (currentStudent.division === undefined) {
            mapSchoolingRegistrationsByStudentId.delete(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID);
        }
```

Reply:
Done